go/logic/topology_recovery.go

```diff
@@ -1479,14 +1479,124 @@ func checkAndRecoverNonWriteableMaster(analysisEntry inst.ReplicationAnalysis, c
 
 // checkAndRecoverLockedSemiSyncMaster
 func checkAndRecoverLockedSemiSyncMaster(analysisEntry inst.ReplicationAnalysis, candidateInstanceKey *inst.InstanceKey, forceInstanceRecovery bool, skipProcesses bool) (recoveryAttempted bool, topologyRecovery *TopologyRecovery, err error) {
-
 	topologyRecovery, err = AttemptRecoveryRegistration(&analysisEntry, true, true)
 	if topologyRecovery == nil {
 		AuditTopologyRecovery(topologyRecovery, fmt.Sprintf("found an active or recent recovery on %+v. Will not issue another RecoverLockedSemiSyncMaster.", analysisEntry.AnalyzedInstanceKey))
 		return false, nil, err
 	}
 
-	return false, nil, nil
+	if !config.Config.EnforceSemiSyncReplicaCount {
+		return false, nil, nil
+	}
+
+	replicas, err := inst.ReadReplicaInstances(&analysisEntry.AnalyzedInstanceKey)
+	if err != nil {
+		return false, topologyRecovery, err
+	}
+	if len(replicas) == 0 {
+		AuditTopologyRecovery(topologyRecovery, fmt.Sprintf("no replicas found for %+v.", analysisEntry.AnalyzedInstanceKey))
+		return false, topologyRecovery, nil
+	}
+
+	reversePriorityReplicas := replicas // TODO sort replicas by priority
```

Comment:
In my opinion, `promotion_rule`s strongly correlate with semi-sync priority. I'd like to suggest dropping the `priority` query column and work  down the promotion rule `must`>`prefer`>`neutal`>`prefer_not`>`must_not`.

As I'm writing this, I also note that region/datacenter play a critical role. The thing is that some people want semi-sync _only within same region/dc_ and some _only outside region/dc_. Which makes it impossible for `orchestrator` to do prioritization.

Still, I feel like the `priority` query puts a burden on the user, and is not trivial to compute. It's e.g. not trivial to generate the query in the config file in a puppet module.

Perhaps two-level sorting: 1st order is user's `priority`, 2nd is promotion-rule?

More ideas are welcome.

Reply:
I like the idea of a two-level priority. I've implemented this idea in the latest PR (though it's not fully tested). It now sorts the replicas in this order:

- priority (I have redefined `SemiSyncEnforced` as an int (lower numbers are higher priorities), which is nice because that actually makes it backwards compatible.
- promotion rule
- instance key (a terrible fallback, but the only way to break a tie)